### ISSUE TITLE
False-positive array_fill()'s start index mutation

### ISSUE THREAD

**Issue by someniatko** [infection/infection#1449]
| Question    | Answer
| ------------| ---------------
| Infection version | 0.20.1
| Test Framework version | PHPUnit 9.3.11
| PHP version | Infection Playground
| Platform    | Infection Playground
| Github Repo | https://infection-php.dev/r/007

Infection mutates such code, by changing (increment/decrement) the first argument of the array_fill, which is responsible for the index of the first element of resulting array. However any indices are destructed by the consequent implode() call.
```php
implode(',', array_fill(0, 2, '(?,?)'))
```

(For a human) it is obvious the array_fill's first parameter does not impact the end result of such code execution, hence it is impossible to write the code and tests in such a way it could pass the mutation testing. I would be very happy if Infection could spot such situation and avoid meaningless mutations.

See https://infection-php.dev/r/007

**Comment by maks-rafalko** [comment]
Not sure it is a bug though. 

If `array_fill()` is used without implode - it **is** possible to write a test to catch this Mutant. But we can't predict all the cases in the code that developers can write.

So, the false-positive here is because of a *combination* of functions used, not because of the original mutator that mutates the first argument of `array_fill`.

The first thoughts here:

* move `0` magic number to a constant, so that Infection does not mutate it: `array_fill(self::SOME_INDEX, self::CONST_FOR_COUNT, '(?,?)')`
* ignore such mutations with [`ignoreSourceCodeByRegex`](https://infection.github.io/guide/usage.html), for example like this https://infection-php.dev/r/4mv or even like this https://infection-php.dev/r/mzd

**Comment by sanmai** [comment]
On one hand `array_fill(0, 2, 'x')` and `array_fill(-1, 2, 'x')` give [the same values](https://3v4l.org/cnEPt), but on the other hand resulting arrays have _different_ keys.

Therefore, this isn't a true "uncatchable" false positive.

**Comment by someniatko** [comment]
> Therefore, this isn't a true "uncatchable" false positive.

@sanmai I am sure that you are not right. If it's not an unchatchable one please then try proving your position by writing a test killing the mutations for my original function with implode. Your argument would only make sense if I returned array_fill()'s output directly. 

Maks Rafalko has a good point that it is combination of functions, not the function itself, and this fact makes a difference. He also have said that infection cannot predict all the cases in code that developers can write. I guess it means it is technically hard and unfeasible to predict such case, because it would probably need to lookup the analyze what does the function operating on the result of array_fill() do, which looks more like a job for Psalm. 

It's sad that I do not have any other option but whitelisting the code chunk, but I guess let it be.

**Comment by sanmai** [comment]
There's nothing to be sad about here. But let's consider that Maks' point about this being a combination of functions. Let's assume you're correct this combination of functions is a _truly uncatchable_ false positive. What shall be our course of action here then?

**Comment by someniatko** [comment]
@sanmai Thank you for being open-minded to such a degree! I've closed the issue because I had realized there is actually nothing to be done on infection side: it would take an enormous analyze work. First complication is if I would first write array_fill()'s output to a variable, and then apply the function on that variable. The second complication is if I would use some user-defined function instead of implode, which also happens to be array keys agnostic.

I therefore think, avoiding the false-positive in a _general_ case is too hard to be considered.
